### PR TITLE
[CPU] V1 support for the CPU backend

### LINKED ISSUE #16056: [Usage]: v1 engine on CPU
### Your current environment

Hi, I am reading this framework recently. I wonder why v1 engine doesn't support CPU yet? Since v1 engine has been built, I couldn't see the obstacle in it. Maybe you could help me sort it out :-) 
Thanks!

### How would you like to use vllm

I want to run inference of a [specific model](put link here). I don't know how to integrate it with vllm.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by bigPYJ1151**
resolve #16056 

Support all features listed in the CPU doc excepts FP8 KV cache.

### Changes

- Add V1 ```CPUWorker``` and ```CPUModelRunner```, derived from ```Worker``` and ```GPUModelRunner``` to reduce code duplication. 
- Add V1 ```TorchSDPABackend``` with compatible interfaces for ```GPUModelRunner```, such as ```reorder_batch``` and ```build```.
- Additional changes in ```GPUModelRunner``` to avoid importing flash-attn explicitly and using default ```nccl``` dist backend.
- Enable CPU tests support the V1 engine.

### CODE REVIEW

**Comment by DarkLight1337** [comment]
Nice, can you also update the V1 User Guide to reflect this support?

File: vllm/v1/worker/gpu_model_runner.py [hunk 1]
```diff
@@ -196,8 +196,10 @@ def __init__(
                 self.vllm_config.compilation_config.cudagraph_capture_sizes))
 
         # Cache the device properties.
-        self.device_properties = torch.cuda.get_device_properties(self.device)
-        self.num_sms = self.device_properties.multi_processor_count
+        if torch.cuda.is_available():
+            self.device_properties = torch.cuda.get_device_properties(
+                self.device)
+            self.num_sms = self.device_properties.multi_processor_count
```
Comment (DarkLight1337) [inline vllm/v1/worker/gpu_model_runner.py:202]: IMO the GPU model runner shouldn't know about the existence of CPU model runner.

Comment (bigPYJ1151) [reply vllm/v1/worker/gpu_model_runner.py:202]: Maybe we should move this to a platform interface or a member function so I can override this?

Comment (DarkLight1337) [reply vllm/v1/worker/gpu_model_runner.py:202]: Yeah let's move this into a member function like `_init_device_properties`

File: vllm/attention/backends/torch_sdpa.py [hunk 2]
```diff
@@ -74,6 +78,59 @@ def copy_blocks(
         PagedAttention.copy_blocks(kv_caches, src_to_dists)
 
 
+class TorchSDPABackendV1:
+    accept_output_buffer: bool = False
+
+    @staticmethod
+    def get_name() -> str:
+        return "TORCH_SDPA"
```
Comment (Isotr0py) [inline vllm/attention/backends/torch_sdpa.py:86]:
```suggestion
        return "TORCH_SDPA_VLLM_V1"
```

File: vllm/v1/worker/cpu_model_runner.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+from vllm.config import VllmConfig
+from vllm.logger import init_logger
+from vllm.model_executor.model_loader import get_model
+from vllm.v1.worker.gpu_model_runner import GPUModelRunner
+
+logger = init_logger(__name__)
+
+
+class CPUModelRunner(GPUModelRunner):
```
Comment (Isotr0py) [inline vllm/v1/worker/cpu_model_runner.py:15]: I think we will also need a `ModelRunnerBase` class for v1 to avoid directly inherit from GPU runner.

File: vllm/attention/backends/torch_sdpa.py [hunk 4]
```diff
@@ -74,6 +78,59 @@ def copy_blocks(
         PagedAttention.copy_blocks(kv_caches, src_to_dists)
 
 
+class TorchSDPABackendV1:
+    accept_output_buffer: bool = False
+
```
Comment (Isotr0py) [inline vllm/attention/backends/torch_sdpa.py:83]: I think we should put the v1 sdpa attention backend at `vllm/v1/attention/backends` to avoid coupling.

File: vllm/attention/backends/torch_sdpa.py [hunk 5]
```diff
@@ -283,6 +343,105 @@ def get_seq_len_block_table_args(
             raise AttributeError(f"Invalid attention type {str(attn_type)}")
 
 
+class TorchSDPAMetadataBuilderV1:
+
```
Comment (Isotr0py) [inline vllm/attention/backends/torch_sdpa.py:347]: ditto.

File: vllm/attention/backends/torch_sdpa.py [hunk 6]
```diff
@@ -74,6 +78,59 @@ def copy_blocks(
...
+class TorchSDPABackendV1:
+    accept_output_buffer: bool = False
+
+    @staticmethod
+    def get_name() -> str:
+        return "TORCH_SDPA"
+
+    @staticmethod
+    def get_impl_cls() -> Type["TorchSDPABackendImpl"]:
+        return TorchSDPABackendImpl
```
Comment (Isotr0py) [inline vllm/attention/backends/torch_sdpa.py:90]: We should add a new AttentionImpl for v1 instead of using the legacy ones, because  prefix caching and chunk prefill are always enabled by default.

**Review by Isotr0py — COMMENTED** [review]
Added some initial comments, PTAL!

File: vllm/v1/worker/cpu_model_runner.py (hunk 3 above)
Comment (bigPYJ1151) [reply vllm/v1/worker/cpu_model_runner.py:15]: Absolutely, maybe it should be done when the GPU runner becomes stable. At least the CPU V1 runner can fully reuse the GPU runner with limited additional changes.

.

File: vllm/attention/backends/torch_sdpa.py (hunk 4 above)
Comment (bigPYJ1151) [reply vllm/attention/backends/torch_sdpa.py]: Good idea :)

File: vllm/attention/backends/torch_sdpa.py (hunk 6 above)
Comment (bigPYJ1151) [reply vllm/attention/backends/torch_sdpa.py]: Yes, ```chunked_prefill``` is always enabled when building metadata for v1. Right now I think the legacy impl is enough for enabling v1 for CPU, and perhaps we can provide a new unified CPU attn impl for multiple attn features in the future.

Comment (Isotr0py) [reply vllm/attention/backends/torch_sdpa.py]: > perhaps we can provide a new unified CPU attn impl for multiple attn features in the future.

Sure! In fact, I also expect Flex Attn can support on CPU backend: https://github.com/vllm-project/vllm/pull/16078. Perhaps we can switch to FlexAttn from SDPA once it lands in the future.

File: vllm/attention/backends/cpu_mla.py [hunk 7]
```diff
@@ -177,7 +177,7 @@ def build(self, seq_lens, query_lens, cuda_graph_pad_size, batch_size):
             seq_lens_tensor=seq_lens_tensor,
             max_query_len=max_query_len,
             max_kv_len=max_kv_len,
```
Comment (robertgshaw2-redhat) [inline vllm/attention/backends/cpu_mla.py:179]: why does this file require changes?

File: vllm/attention/backends/torch_sdpa.py [hunk 8]
```diff
@@ -86,10 +86,13 @@ class TorchSDPAMetadata(AttentionMetadata, PagedAttentionMetadata):
     # For chunked prefill only
     max_query_len: Optional[int] = None
     max_kv_len: Optional[int] = None
-    query_start_loc: Optional[torch.Tensor] = None
```
Comment (robertgshaw2-redhat) [inline vllm/attention/backends/torch_sdpa.py:89]: why does this file require changes?

File: vllm/v1/attention/backends/cpu_attn.py [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+        num_blocks: int,
+        block_size: int,
+        num_kv_heads: int,
+        head_size: int,
+    ) -> tuple[int, ...]:
+        return PagedAttention.get_kv_cache_shape(num_blocks, block_size,
+                                                 num_kv_heads, head_size)
+
+    @staticmethod
+    def swap_blocks(
```
Comment (robertgshaw2-redhat) [inline vllm/v1/attention/backends/cpu_attn.py:49]: I dont think swap_blocks or copy_blocks are needed for V1

File: vllm/attention/backends/torch_sdpa.py (hunk 8 above)
Comment (bigPYJ1151) [reply vllm/attention/backends/torch_sdpa.py:89]: It is a naming conflict. The V1 model runner use ```query_start_loc``` for logits indexing specially, contains all tokens in a batch. But in ```torch_sdpa```, ```query_start_loc``` contains prefill tokens only, so rename it to ```prefill_query_start_loc```.

File: vllm/v1/worker/cpu_model_runner.py [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+
+        self.seq_start_loc_cpu = torch.zeros(
+            self.max_num_reqs + 1,
+            dtype=torch.int32,
+            device="cpu",
+        )
+        self.seq_start_loc_np = self.seq_start_loc_cpu.numpy()
+
+    def _postprocess_tenosrs(self) -> None:
+        # Note: replace device tensors with cpu tensors
```
Comment (NickLucche) [inline vllm/v1/worker/cpu_model_runner.py:38]: This is a bit hacky. Have you considered just duplicating GPUModelRunner while cutting out all unsupported features or unnecessary device/cpu double tensors?
We have done something similar for TPU https://github.com/vllm-project/vllm/blob/main/vllm/v1/worker/tpu_model_runner.py.

It's uglier to the eyes but it's easier to control until things are more stable.

**Review by NickLucche — CHANGES_REQUESTED** [review]
Thanks for the contribution!

File: vllm/platforms/cpu.py [hunk 11]
```diff
@@ -65,6 +68,12 @@ def check_and_update_config(cls, vllm_config: VllmConfig) -> None:
         if not model_config.enforce_eager:
             model_config.enforce_eager = True
 
+        if model_config.enable_sleep_mode:
+            logger.warning("sleep mode is not supported on CPU, disable it.")
+            model_config.enable_sleep_mode = False
+
+        model_config.disable_cascade_attn = True
```
Comment (simon-mo) [inline vllm/platforms/cpu.py:75]: This is new?

**Comment by xuechendi** [comment]
@bigPYJ1151 , we had some discussion regards this PR in HPU channel, Simon requested to list changes in the PR description. Thanks

File: vllm/v1/worker/cpu_model_runner.py (hunk 10 above)
Comment (bigPYJ1151) [reply vllm/v1/worker/cpu_model_runner.py:38]: Exactly it is a bit hacky. We actually did the way in V0, it did well to cut out some hardcoded unsupported features for CPU such as CUDAGraph, FlashAttnBackend. But eventually the V0 CPU runner diverged with the GPU runner and can't  leverage some new features and refactors directly.

Compared with V0, V1 model runner has better designs in abstraction, like unified input data management, vLLM compilation, attention backend builder and reorder, .etc. These abstraction allows the CPU model runner be more compatible with the GPU. Moreover, I think the V1 model runner has became relatively stable as almost no changes required when rebasing this PR.

So I would prefer to totally reuse the GPU model runner for now, even with some hacky. If some day we have to decouple them, I think it will not be very difficult.

File: vllm/platforms/cpu.py (hunk 11 above)
Comment (bigPYJ1151) [reply vllm/platforms/cpu.py:75]: Yes I think the cascade attn is only supported in the flash attn backend so I disable it here.

I noticed ```support_sleep_mode``` has became a platform attribute, remove the checking here.

File: vllm/v1/worker/cpu_model_runner.py (hunk 10 above)
Comment (NickLucche) [reply vllm/v1/worker/cpu_model_runner.py:38]: While I definitely see your point and the strengths of leveraging an hierarchical structure, I am not 100% sure there won't be quite recurring hiccups to fix the runner as an unintended consequence of some other unrelated PR.

Still, I am not against this implementation, perhaps we could a better job at writing platform-aware code in the other parts of the code.

Eg (not for this PR) abstracting away those cpu/device tensors in a base class so that we don't have to resort to this workaround here.

**Review by NickLucche — CHANGES_REQUESTED** [review]
May I ask what kind of features do you expect to support or not support with this PR? 
Eg is MM or spec decoding working fine?

**Comment by bigPYJ1151** [comment]
> May I ask what kind of features do you expect to support or not support with this PR? Eg is MM or spec decoding working fine?

Right now the multi-modal is supported and I have verified it initially. For spec decoding I disabled it as the V0 CPU also doesn't support. In general I expect to migrate all supported features in V0 CPU to V1.

File: vllm/attention/backends/torch_sdpa.py (hunk 8 above)
Comment (simon-mo) [reply vllm/attention/backends/torch_sdpa.py:89]: I would try to avoid the renaming in torch sdpa instead of the global change

**Review by jikunshang — APPROVED** [review]

File: vllm/v1/worker/gpu_model_runner.py [hunk 12]
```diff
@@ -310,6 +308,15 @@ def __init__(
                                         pin_memory=self.pin_memory)
         self.seq_lens_np = self.seq_lens_cpu.numpy()
 
+    def _init_device_properties(self) -> None:
+        """Initialize attributes from torch.cuda.get_device_properties
+        """
+        self.device_properties = torch.cuda.get_device_properties(self.device)
+        self.num_sms = self.device_properties.multi_processor_count
+
+    def _sync_device(self) -> None:
```
Comment (simon-mo) [inline vllm/v1/worker/gpu_model_runner.py:317]: add a comment saying this is required for model runner override

**Review by simon-mo — APPROVED** [review]

**Comment by xuechendi** [comment]
@bigPYJ1151 , please take a look of failed CI:
image

**Comment by bigPYJ1151** [comment]
Hi @simon-mo, all required tests became green. We enabled CPU V1 as default and all CPU tests compatible with the V1 engine passed. https://buildkite.com/vllm/ci/builds/21328#0197361e-3946-4915-953d-7a8baeae1137

I think this PR is ready to merge, please take a look, thanks :)
